File: scanner.py
"""
scanner.py
----------
The scanning engine. Contains no GUI code so it can be unit tested or
reused independently of CustomTkinter. Runs a TCP Connect scan across
a pool of worker threads and reports progress/results via callbacks,
so the GUI thread is never blocked.

Only intended for use against hosts you own or are explicitly
authorized to test.
"""

from __future__ import annotations

import socket
import threading
import time
from concurrent.futures import ThreadPoolExecutor, Future
from dataclasses import dataclass, field
from typing import Callable, Optional

from services import get_service
from utils import Stopwatch, now_str


@dataclass
class ScanResult:
    port: int
    status: str  # "open" | "closed" | "filtered"
    service: str
    banner: Optional[str] = None
    response_ms: float = 0.0
    scanned_at: str = field(default_factory=now_str)


# Callback signatures used to keep the GUI decoupled from the engine.
ProgressCallback = Callable[[int, int, float], None]  # done, total, elapsed
ResultCallback = Callable[[ScanResult], None]
DoneCallback = Callable[[list[ScanResult], float, bool], None]  # results, duration, was_stopped


class PortScanner:
    """
    Threaded TCP Connect port scanner.

    Usage:
        scanner = PortScanner(target, start_port, end_port,
                               timeout=0.5, max_threads=100,
                               on_progress=..., on_result=..., on_done=...)
        scanner.start()
        ...
        scanner.stop()   # optional, graceful cancellation
    """

    def __init__(
        self,
        target_ip: str,
        start_port: int,
        end_port: int,
        timeout: float = 0.5,
        max_threads: int = 100,
        grab_banner: bool = True,
        on_progress: Optional[ProgressCallback] = None,
        on_result: Optional[ResultCallback] = None,
        on_done: Optional[DoneCallback] = None,
    ) -> None:
        self.target_ip = target_ip
        self.start_port = start_port
        self.end_port = end_port
        self.timeout = timeout
        self.max_threads = max_threads
        self.grab_banner = grab_banner

        self.on_progress = on_progress
        self.on_result = on_result
        self.on_done = on_done

        self._stop_event = threading.Event()
        self._lock = threading.Lock()
        self._completed = 0
        self._results: list[ScanResult] = []
        self._stopwatch = Stopwatch()
        self._executor: Optional[ThreadPoolExecutor] = None
        self._runner_thread: Optional[threading.Thread] = None

    @property
    def total_ports(self) -> int:
        return self.end_port - self.start_port + 1

    def start(self) -> None:
        """Kick off the scan on a background thread (non-blocking)."""
        self._stop_event.clear()
        self._completed = 0
        self._results = []
        self._stopwatch.start()
        self._runner_thread = threading.Thread(target=self._run, daemon=True)
        self._runner_thread.start()

    def stop(self) -> None:
        """Request graceful cancellation. In-flight sockets finish, no new ones start."""
        self._stop_event.set()

    def is_stopped(self) -> bool:
        return self._stop_event.is_set()

    # -- internals ----------------------------------------------------

    def _scan_port(self, port: int) -> ScanResult:
        start = time.perf_counter()
        service = get_service(port)
        banner = None
        status = "closed"

        try:
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
                sock.settimeout(self.timeout)
                result_code = sock.connect_ex((self.target_ip, port))
                if result_code == 0:
                    status = "open"
                    if self.grab_banner:
                        banner = self._try_grab_banner(sock)
                else:
                    status = "closed"
        except socket.timeout:
            status = "filtered"
        except OSError:
            status = "filtered"

        elapsed_ms = (time.perf_counter() - start) * 1000
        return ScanResult(
            port=port,
            status=status,
            service=service,
            banner=banner,
            response_ms=round(elapsed_ms, 2),
        )

    @staticmethod
    def _try_grab_banner(sock: socket.socket) -> Optional[str]:
        """Best-effort banner grab; never raises, never blocks for long."""
        try:
            sock.settimeout(0.4)
            data = sock.recv(128)
            if data:
                return data.decode(errors="replace").strip()[:120]
        except (socket.timeout, OSError):
            return None
        return None

    def _run(self) -> None:
        ports = range(self.start_port, self.end_port + 1)
        was_stopped = False

        with ThreadPoolExecutor(max_workers=self.max_threads) as executor:
            self._executor = executor
            futures: dict[Future, int] = {}
            for port in ports:
                if self._stop_event.is_set():
                    was_stopped = True
                    break
                futures[executor.submit(self._scan_port, port)] = port

            for future in futures:
                if self._stop_event.is_set():
                    was_stopped = True
                # Still collect results already in-flight/completed.
                try:
                    result = future.result()
                except Exception:
                    continue

                with self._lock:
                    self._completed += 1
                    self._results.append(result)
                    done = self._completed

                if self.on_result:
                    self.on_result(result)
                if self.on_progress:
                    self.on_progress(done, self.total_ports, self._stopwatch.elapsed())

        elapsed = self._stopwatch.elapsed() or 0.0
        self._stopwatch.stop()
        # Sort results by port for a stable, readable output order.
        self._results.sort(key=lambda r: r.port)
        if self.on_done:
            self.on_done(self._results, elapsed, was_stopped)

File: src/scanner.ts
import net from "node:net";
import { getService } from "./services.js";
import { Stopwatch, nowStr } from "./utils.js";

export interface ScanResult {
  port: number;
  status: "open" | "closed" | "filtered";
  service: string;
  banner: string | null;
  response_ms: number;
  scanned_at: string;
}

export type ProgressCallback = (done: number, total: number, elapsed: number) => void;
export type ResultCallback = (result: ScanResult) => void;
export type DoneCallback = (results: ScanResult[], duration: number, stopped: boolean) => void;

export class PortScanner {
  public targetIp: string;
  public ports: number[];
  public timeout: number;
  public maxThreads: number;
  public grabBanner: boolean;

  public onProgress?: ProgressCallback;
  public onResult?: ResultCallback;
  public onDone?: DoneCallback;

  private _stopped: boolean = false;
  private _completed: number = 0;
  private _results: ScanResult[] = [];
  private _stopwatch: Stopwatch = new Stopwatch();

  constructor(options: {
    targetIp: string;
    ports?: number[];
    startPort?: number;
    endPort?: number;
    timeout?: number;
    maxThreads?: number;
    grabBanner?: boolean;
    onProgress?: ProgressCallback;
    onResult?: ResultCallback;
    onDone?: DoneCallback;
  }) {
    this.targetIp = options.targetIp;

    if (options.ports && options.ports.length > 0) {
      this.ports = Array.from(new Set(options.ports)).sort((a, b) => a - b);
    } else {
      const s = options.startPort ?? 1;
      const e = options.endPort ?? 1024;
      const list: number[] = [];
      for (let p = s; p <= e; p++) {
        list.push(p);
      }
      this.ports = list;
    }

    this.timeout = options.timeout ?? 0.5;
    this.maxThreads = options.maxThreads ?? 100;
    this.grabBanner = options.grabBanner ?? true;

    this.onProgress = options.onProgress;
    this.onResult = options.onResult;
    this.onDone = options.onDone;
  }

  get totalPorts(): number {
    return this.ports.length;
  }

  public start(): void {
    this._stopped = false;
    this._completed = 0;
    this._results = [];
    this._stopwatch.start();
    this.runPool().catch((err) => {
      console.error("Error in port scanner worker pool:", err);
    });
  }

  public stop(): void {
    this._stopped = true;
  }

  public isStopped(): boolean {
    return this._stopped;
  }

  private scanPort(port: number): Promise<ScanResult> {
    const startTime = performance.now();
    const service = getService(port);
    const timeoutMs = Math.max(50, Math.round(this.timeout * 1000));

    return new Promise<ScanResult>((resolve) => {
      let isSettled = false;
      const socket = new net.Socket();
      let timer: NodeJS.Timeout | null = null;
      let bannerTimer: NodeJS.Timeout | null = null;
      let banner: string | null = null;

      const cleanup = () => {
        if (timer) {
          clearTimeout(timer);
          timer = null;
        }
        if (bannerTimer) {
          clearTimeout(bannerTimer);
          bannerTimer = null;
        }
        socket.removeAllListeners();
        socket.destroy();
      };

      const finish = (status: "open" | "closed" | "filtered") => {
        if (isSettled) return;
        isSettled = true;
        const elapsedMs = performance.now() - startTime;
        cleanup();
        resolve({
          port,
          status,
          service,
          banner,
          response_ms: Math.round(elapsedMs * 10) / 10,
          scanned_at: nowStr(),
        });
      };

      // Connect timeout timer
      timer = setTimeout(() => {
        finish("filtered");
      }, timeoutMs);

      socket.setTimeout(timeoutMs);

      // Attempt TCP 3-way handshake
      socket.connect(
        {
          port,
          host: this.targetIp,
        },
        () => {
          // Connected successfully => Port is definitely OPEN!
          // Crucial: Clear the connect timeout immediately so it never falsely marks as filtered
          if (timer) {
            clearTimeout(timer);
            timer = null;
          }

          if (!this.grabBanner) {
            finish("open");
            return;
          }

          // Banner grab window
          bannerTimer = setTimeout(() => {
            finish("open");
          }, 150);

          socket.on("data", (chunk: Buffer) => {
            const raw = chunk
              .toString("utf-8")
              .replace(/[\r\n\x00-\x1F\x7F]+/g, " ")
              .trim();
            if (raw) {
              banner = raw.slice(0, 100);
            }
            finish("open");
          });

          socket.on("error", () => {
            finish("open");
          });

          socket.on("close", () => {
            finish("open");
          });

          // Send lightweight probe based on common services
          try {
            if (port === 80 || port === 8080 || port === 3000 || port === 5000 || port === 8000) {
              socket.write(`HEAD / HTTP/1.0\r\nHost: ${this.targetIp}\r\n\r\n`);
            } else if (port === 443 || port === 8443) {
              // SSL/TLS will respond or disconnect; already open
            } else {
              socket.write("\r\n");
            }
          } catch {
            // ignore probe write errors
          }
        }
      );

      socket.on("timeout", () => {
        finish("filtered");
      });

      socket.on("error", (err: any) => {
        const code = err?.code || "";
        if (code === "ECONNREFUSED") {
          // RST received => Host is alive, port is closed
          finish("closed");
        } else if (code === "ECONNRESET" || code === "EPIPE") {
          // Connection accepted then reset => open
          finish("open");
        } else if (
          code === "ETIMEDOUT" ||
          code === "EHOSTUNREACH" ||
          code === "ENETUNREACH" ||
          code === "EHOSTDOWN" ||
          code === "ENETDOWN" ||
          code === "ENOTFOUND"
        ) {
          // Firewall dropped packets or network unreachable
          finish("filtered");
        } else {
          // General socket failure
          finish("closed");
        }
      });
    });
  }

  private async runPool(): Promise<void> {
    const ports = this.ports;
    let currentIndex = 0;
    const workerCount = Math.min(this.maxThreads, ports.length);

    const worker = async () => {
      while (currentIndex < ports.length && !this._stopped) {
        const portIndex = currentIndex++;
        const port = ports[portIndex];
        const result = await this.scanPort(port);

        this._completed++;
        this._results.push(result);

        if (this.onResult) {
          this.onResult(result);
        }
        if (this.onProgress) {
          this.onProgress(this._completed, this.totalPorts, this._stopwatch.elapsed());
        }
      }
    };

    const workers: Promise<void>[] = [];
    for (let i = 0; i < workerCount; i++) {
      workers.push(worker());
    }

    await Promise.all(workers);

    const elapsed = this._stopwatch.elapsed() || 0.0;
    this._stopwatch.stop();

    // Stable sort by port
    this._results.sort((a, b) => a.port - b.port);

    if (this.onDone) {
      this.onDone(this._results, elapsed, this._stopped);
    }
  }
}
